include/robot_interfaces/robot_backend.hpp

```diff
@@ -128,12 +136,12 @@ class RobotBackend
 
         // wait until first desired_action was received
         // ----------------------------
-        while (!destructor_was_called_ &&
+        while (!should_stop_ &&
                !robot_data_->desired_action->wait_for_timeindex(0, 0.1))
         {
         }
```

Comment:
i think actually the problem could be solved by just moving the block above into the for loop below, i.e. put

Reply:
Just moving the loop would not be enough but something similar should work.

@vincentberenz Let's talk about this on Monday.

Reply:
Not tested but I think something like the following should work:
```c++
while (!robot_data_->desired_action->wait_for_timeindex(t, 0.01)) {
    if (destructor_was_called_) {
        return;
    }
}

Action desired_action = (*robot_data_->desired_action)[t];
...
```
Note that I chose the timeout duration of 10ms rather arbitrarily, not sure what would be a good value.

Reply:
yes, i think that should work. unlike what i said, i think we also need to leave the check before the loop. otherwise, i think we might have a similar problem in line 177 in the first iteration...

since this seems to be a quite common problem, i am starting to think that the best thing to do would be what @luator and me discussed: add a kill() function to Timeseries, which will make all the functions which are currently waiting for the next element throw an exception.

then we can simply kill all the timeseries in the destructor of the classes, and we immediately see if we did not handle the exception somewhere, and the hanging cannot happen anymore.

what do you think?

Reply:
@luator @mwuethri I implemented the solution suggested by Felix, and it worked, at least on the demo. I commited and push. I do not know the code enough to have an interesting opinion, but a kill function sounds useful. But in the meantime merging this would help, this is blocking for me.